src/util/codec/mysql/decimal.rs

```diff
@@ -11,770 +11,2008 @@
-        }
-
-        let to_add = exp - self.exp;
-        let coeff = if to_add > 0 {
-            let divider = BigInt::from(num::pow(BigUint::from(10u8), to_add as usize));
-            let (mut coeff, rem) = self.coeff.div_rem(&divider);
-            if rem >= divider / BigInt::from(2) {
-                coeff = coeff + BigInt::one();
-            }
-            coeff
-        } else {
-            self.coeff.clone() * BigInt::from(num::pow(BigUint::from(10u8), -to_add as usize))
-        };
+/// add lhs to rhs.
+fn do_add(mut lhs: Decimal, mut rhs: Decimal) -> Res<Decimal> {
+    let (mut l_int_word_cnt, mut l_frac_word_cnt) = (word_cnt!(lhs.int_cnt),
+                                                     word_cnt!(lhs.frac_cnt));
+    let (mut r_int_word_cnt, mut r_frac_word_cnt) = (word_cnt!(rhs.int_cnt),
+                                                     word_cnt!(rhs.frac_cnt));
+    let (mut int_word_to, frac_word_to) = (cmp::max(l_int_word_cnt, r_int_word_cnt),
+                                           cmp::max(l_frac_word_cnt, r_frac_word_cnt));
+    let x = if l_int_word_cnt > r_int_word_cnt {
+        lhs.word_buf[0]
+    } else if l_int_word_cnt < r_int_word_cnt {
+        rhs.word_buf[0]
+    } else {
+        lhs.word_buf[0] + rhs.word_buf[0]
+    };
+    if x > WORD_MAX - 1 {
+        int_word_to += 1;
+    }
+    let res = fix_word_cnt_err(int_word_to, frac_word_to, WORD_BUF_LEN);
+    if res.is_overflow() {
+        return Res::Overflow(max_decimal(WORD_BUF_LEN * DIGITS_PER_WORD, 0));
+    }
+    let (int_word_to, frac_word_to) = res.clone().unwrap();
+    let mut idx_to = (int_word_to + frac_word_to) as usize;
+    let mut res = res.map(|_| {
+        Decimal::new(int_word_to * DIGITS_PER_WORD,
+                     cmp::max(lhs.frac_cnt, rhs.frac_cnt),
+                     lhs.negative)
+    });
+    res.word_buf[0] = 0;
+    if !res.is_ok() {
+        res.frac_cnt = cmp::min(frac_word_to * DIGITS_PER_WORD, res.frac_cnt);
+        l_frac_word_cnt = cmp::min(frac_word_to, l_frac_word_cnt);
+        r_frac_word_cnt = cmp::min(r_frac_word_cnt, frac_word_to);
+        l_int_word_cnt = cmp::min(l_int_word_cnt, int_word_to);
+        r_int_word_cnt = cmp::min(r_int_word_cnt, int_word_to);
+    }
+    let (mut l_idx, mut r_idx, l_stop, r_stop, exchanged);
+    if l_frac_word_cnt > r_frac_word_cnt {
+        l_idx = (l_int_word_cnt + l_frac_word_cnt) as usize;
+        l_stop = (l_int_word_cnt + r_frac_word_cnt) as usize;
+        r_idx = (r_int_word_cnt + r_frac_word_cnt) as usize;
+        r_stop = l_int_word_cnt.saturating_sub(r_int_word_cnt) as usize;
+        exchanged = false;
+    } else {
+        l_idx = (r_int_word_cnt + r_frac_word_cnt) as usize;
+        l_stop = (r_int_word_cnt + l_frac_word_cnt) as usize;
+        r_idx = (l_int_word_cnt + l_frac_word_cnt) as usize;
+        r_stop = r_int_word_cnt.saturating_sub(l_int_word_cnt) as usize;
+        mem::swap(&mut lhs, &mut rhs);
+        exchanged = true;
```

Comment:
When using temporary variable, we will have to clone it or deal with lifetime issues, both will lead to codes that is not very clean.